subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CompositeCallInterceptor.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath.intercept;
+
+import org.gradle.internal.instrumentation.api.types.BytecodeInterceptorType;
+
+import javax.annotation.Nullable;
+
+public class CompositeCallInterceptor extends CallInterceptor implements SignatureAwareCallInterceptor, PropertyAwareCallInterceptor {
+
+    private final CallInterceptor first;
+    private final CallInterceptor second;
+
+    public CompositeCallInterceptor(CallInterceptor first, CallInterceptor second) {
+        this.first = first;
+        this.second = second;
+    }
+
+    @Override
+    public Object doIntercept(Invocation invocation, String consumer) throws Throwable {
+        return first.doIntercept(new Invocation() {
+            @Override
+            public Object getReceiver() {
+                return invocation.getReceiver();
+            }
+
+            @Override
+            public int getArgsCount() {
+                return invocation.getArgsCount();
+            }
+
+            @Override
+            public Object getArgument(int pos) {
+                return invocation.getArgument(pos);
+            }
+
+            @Override
+            public Object callOriginal() throws Throwable {
+                return second.doIntercept(invocation, consumer);
+            }
+        }, consumer);
+    }
+
+    @Override
+    public BytecodeInterceptorType getType() {
+        throw new UnsupportedOperationException("Calling CompositeCallInterceptor.getType() is not supported");
+    }
```

Comment:
:x: This sounds wrong. We either don't need this method, in which case we should delete it. Or if we do, we need to return something meaningful here. How does `BytecodeInterceptorFilter` deal with the thrown exception here?

If we do need this, maybe the right approach is to check if both interceptors have the same type, and then return that; and throw early if they are different. But my suspicion is that we don't need this.

Another option is that we might only need this method in some specific contexts, and `CompositeCallInterceptor` does not actually need to extend `CallInterceptor`. If we go with this option, then we shouldn't call this class a `CallInterceptor` either.

Yet another option is to abstract out some new interface from `CallInterceptor` that does not include `getType()` and make `CallInterceptor` and this class implement that interface instead.

Reply:
Yes, this method is not needed at this point anymore.

Reply:
Maybe I can move this method to some other interface 🤔

Reply:
I changed the hierarchy:
1. I renamed `BytecodeInterceptor` to [FilterableBytecodeInterceptor](https://github.com/gradle/gradle/pull/28746/files#diff-31dec7b0f593d1b2fe0afafee71495a46bfba92355951216bfae00ad49f29f85)
2. I changed [CallInterceptor](https://github.com/gradle/gradle/pull/28746/files#diff-a90a7627da28619007d32178d76736f5f2be27919f20baeb0912ee8764ce8695) from abstract class to interface
3. I introduced [AbstractCallInterceptor](https://github.com/gradle/gradle/pull/28746/files#diff-427161dc0d489a1f844d3d6fd3dcaed1532bc520f881dc52a6a09f4e9c211eb2) that implements `CallInterceptor` with previous implementation
4. I introduced [FilterableCallInterceptor](https://github.com/gradle/gradle/pull/28746/files#diff-9944eaee9d069fcc52f659adfefbae82cd6977c624df7acd13cae7335c005ecf) interface that extends `CallInterceptor` and `FilterableBytecodeInterceptor`
5. All interceptors that are used during filtering are now of type `FilterableCallInterceptor`